### PR TITLE
AT+CIPSERVER & co.

### PR DESCRIPTION
**Pull request by jandrassy**

### CODE REVIEW

**Comment by jandrassy** [comment]
the standard AT firmware supports only one server. we support up to 5 with same AT commands.
In standard AT firmware 1.7 executing `AT+CIPSERVER=1,<port>` again even if the port is different prints `no change` and `OK`.
In standard AT firmware 1.7 executing `AT+CIPSERVER=0` stops the one server. Here it stops the first one. executing `AT+CIPSERVER=0,<port>` is valid for AT 1.7 according to doc. Here it stops the server listening one that port.
I decided to use CIPSERVERMAXCONN and CIPSTO as global settings, not per server, for now. But my library has it as per server setting now.

I modified the WiFiEspAT library for multi server if WIFIESPAT_MULTISERVER is defined.

**Comment by JiriBilek** [comment]
Juraj, thanks for the contribution.
I have never used AT-commands for creating a server, could you please write here instructions for me how to test it?
Thanks.

**Comment by jandrassy** [comment]
> Juraj, thanks for the contribution. I have never used AT-commands for creating a server, could you please write here instructions for me how to test it? Thanks.

You can try it with server examples of the WiFiEspAT library.
It is only the one command like `AT+CIPSERVER=1,80`. Then new client connections are reported with standard `<linkId>,CONNECT` and `+IPD` if data arrive.

File: ESP_ATMod/ESP_ATMod.ino [hunk 1]
```diff
@@ -134,6 +138,8 @@ uint16_t gsCipSslSize = 16384;		// command AT+CIPSSLSIZE
 bool gsSTNPEnabled = true;			// command AT+CIPSNTPCFG
 int8_t gsSTNPTimezone = 0;			// command AT+CIPSNTPCFG
 String gsSNTPServer[3];				// command AT+CIPSNTPCFG
+uint8_t gsServersMaxConn = 5;			// command AT+CIPSERVERMAXCONN
+uint32_t gsServerConnTimeout = 180;	// command AT+CIPSSTO
```
Comment (JiriBilek) [inline ESP_ATMod/ESP_ATMod.ino:142]: The timeout value is given in milliseconds, therefore this value is effectively 0. I think it should be 180*1000.

File: ESP_ATMod/command.cpp [hunk 2]
```diff
@@ -1643,6 +1669,165 @@ void cmd_AT_CIPMUX()
...
+	do
+	{
+		if (inputBuffer[offset] != '=')
+			break;
+
+		++offset;
+		stop = inputBuffer[offset] == '0';
+		++offset;
+
+		if (inputBufferCnt > offset)
```
Comment (JiriBilek) [inline ESP_ATMod/command.cpp:1698]: I think it should be offset + 2, there is a CR/LF at the end of the buffer

File: ESP_ATMod/command.cpp [hunk 3]
```diff
@@ -1643,6 +1669,165 @@ void cmd_AT_CIPMUX()
...
+		{
+			if (inputBuffer[offset] != ',')
+				break;
+
+			++offset;
+			error = 2;
+
+			if (!readNumber(inputBuffer, offset, port) || port > 65535)
+				break;
```
Comment (JiriBilek) [inline ESP_ATMod/command.cpp:1707]: There is no end of buffer checking, so I can type `AT+CIPSERVER=1,8080,ThisIsAnError` and it works. Should be ERROR,, no extra parameters allowed.

**Review by JiriBilek — CHANGES_REQUESTED** [review]
Hi, thanks for the code. I tested it and found several issues. Will continue testing asap.
Thanks, Jiri

File: ESP_ATMod/command.cpp [hunk 4]
```diff
@@ -1643,6 +1669,165 @@ void cmd_AT_CIPMUX()
...
+	bool stop = false;
+	uint32_t port = 0;
+
+	do
+	{
+		if (inputBuffer[offset] != '=')
+			break;
+
+		++offset;
+		stop = inputBuffer[offset] == '0';
```
Comment (JiriBilek) [inline ESP_ATMod/command.cpp:1695]: As far as I understand, the first parameter has 2 allowed values only: `0` and `1`. I'd prefer checking the parameter fully, i.e. throw an error when it is not either 0 or 1.

File: ESP_ATMod/ESP_ATMod.ino [hunk 5]
```diff
@@ -41,6 +41,7 @@
  * 0.3.0: Stored certificates in filesystem with LittleFS
  * 0.3.1: AT+CWLAP, AT+CWLAPOPT and 'busy p...' fix
  * 0.3.2: AT+CWHOSTNAME
+ * 0.3.3: AT+CIPSERVER [J.A]
```
Comment (JiriBilek) [inline ESP_ATMod/ESP_ATMod.ino:44]: Please add here all the newly implemented commands' names

File: ESP_ATMod/ESP_ATMod.ino (hunk 1 above)
Comment (jandrassy) [reply ESP_ATMod/ESP_ATMod.ino:142]: according to AT commnads reference the CIPSTO  value is in seconds

File: ESP_ATMod/command.cpp (hunk 2 above)
Comment (jandrassy) [reply ESP_ATMod/command.cpp:1698]: next character can my a `,` if there is a second parameter (port)

File: ESP_ATMod/ESP_ATMod.ino (hunk 1 above)
Comment (JiriBilek) [reply ESP_ATMod/ESP_ATMod.ino:142]: Yes, it is, but the value in gsServerConnTimeout is in milliseconds, AFAIK

File: ESP_ATMod/command.cpp (hunk 2 above)
Comment (JiriBilek) [reply ESP_ATMod/command.cpp:1698]: Juraj, it really doesn't work. Try AT+CIPSERVER=1, it gives error

File: ESP_ATMod/command.cpp (hunk 3 above)
Comment (jandrassy) [reply ESP_ATMod/command.cpp:1707]: sorry but I don't see such check in other commands processing. I did it by example, not my way.

File: ESP_ATMod/command.cpp (hunk 2 above)
Comment (jandrassy) [reply ESP_ATMod/command.cpp:1698]: it should give ERROR. syntax is `AT+CIPSERVER=<mode>[,<port>]`

File: ESP_ATMod/command.cpp (hunk 3 above)
Comment (jandrassy) [reply ESP_ATMod/command.cpp:1707]: ok I see that the check is `&& inputBufferCnt == offset + 2`

File: ESP_ATMod/command.cpp (hunk 2 above)
Comment (JiriBilek) [reply ESP_ATMod/command.cpp:1698]: Sorry, I hate arguing but it should not. I tested it on AT command processor for ESP32-C3. The command `AT+CIPSERVER=1` is perfectly valid and opens port 333.

Comment (jandrassy) [reply ESP_ATMod/command.cpp:1700]: yes `AT+CIPSERVER=1` is valid, `AT+CIPSERVER=1,` is not. byt please don't check with AT2. the syntax is not the same on some commands.  I expected `inputBuffer` to be without CRLF.

Comment (JiriBilek) [reply ESP_ATMod/command.cpp]: I see the misunderstanding now, sorry. I should have marked the text in my contribution above. I agree with your last reply.

**Review by JiriBilek — APPROVED** [review]
Looks good, thanks

**Pull request merged**